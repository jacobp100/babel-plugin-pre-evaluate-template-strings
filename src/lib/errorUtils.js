/* @flow */

import { SourceMapConsumer } from 'source-map';
import codeFrame from 'babel-code-frame';
import errorStackParser from 'error-stack-parser';

type Frame = {
  functionName: string,
  lineNumber: number,
  columnNumber: number,
  fileName: string,
  source: string,
};

type EnhancedFrame = Frame & { originalSource: string };

export function getFramesFromStack(
  error: Error,
  findLastFrame?: (frames: Frame[]) => number
): Frame[] {
  const allFrames = errorStackParser.parse(error);
  const lastMeaningfulFrame = findLastFrame
    ? findLastFrame(allFrames)
    : allFrames.length - 1;
  return allFrames.slice(0, lastMeaningfulFrame + 1);
}

export function enhanceFrames(
  frames: Frame[],
  modulesCache: { [key: string]: Object }
): EnhancedFrame[] {
  return frames.map(frame => {
    const consumer = SourceMapConsumer(modulesCache[frame.fileName].sourceMap);

    const originalPosition = consumer.originalPositionFor({
      line: frame.lineNumber,
      column: frame.columnNumber,
    });

    return {
      ...frame,
      lineNumber:
        originalPosition.line !== null
          ? originalPosition.line
          : frame.lineNumber,
      columnNumber:
        originalPosition.column !== null
          ? originalPosition.column
          : frame.columnNumber,
      originalSource: consumer.sourcesContent[0],
    };
  });
}

export function buildCodeFrameError(
  error: Error,
  frames: EnhancedFrame[]
): Error {
  const firstFrame = frames && frames[0];

  if (!firstFrame) {
    return error;
  }

  const codeFrameString = codeFrame(
    firstFrame.originalSource,
    firstFrame.lineNumber,
    firstFrame.columnNumber,
    {
      highlightCode: true,
      linesAbove: 4,
      /* istanbul ignore next line */
      forceColor: !!process.env.CI,
    }
  );

  const callStack = frames
    .map(
      frame => `at ${frame.fileName}:${frame.lineNumber}:${frame.columnNumber}`
    )
    .join('\n');

  error.stack = `Error: ${error.message}\n${codeFrameString}\n${callStack}`;
  return error;
}
