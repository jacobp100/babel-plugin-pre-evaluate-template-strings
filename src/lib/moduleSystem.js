/* @flow */

/*
 * ==============================================
 * To avoid leakage from evaled code to module cache in current context,
 * for example with `babel-register` we provide our custom module system.
 * It's designed to mimic the native node one, with the exception being
 * that we can transpile every module by default and store source maps
 * for it. As a result we can provide correct error stacktrace and
 * enhanced errors.
 * ==============================================
 */

import path from 'path';
import fs from 'fs';
import vm from 'vm';
// $FlowFixMe
import NativeModule from 'module';
import * as babel from 'babel-core';

import {
  buildCodeFrameError,
  enhanceFrames,
  getFramesFromStack,
} from './errorUtils';

type Exports = any;

/**
 * Separate cache for evaled modules, so that we don't have to worry about
 * babel-register leaking to main context.
 */
const modulesCache: { [key: string]: Module } = {};

/**
 * Resolve module id (and filename) relatively to parent module if specified.
 */
function resolveModuleId(moduleId: string, parent: ?Module): string {
  return NativeModule._resolveFilename(moduleId, parent, false);
}

/**
 * Create module instance and store it in cache.
 */
export function instantiateModule(
  code: string,
  filename: string,
  parent: ?Module
): Module {
  const moduleInstance = new Module(filename, parent);
  // Store it in cache at this point with loaded: false flag, otherwise
  // we would end up in infinite loop, with cycle dependencies.
  modulesCache[filename] = moduleInstance;

  moduleInstance._compile(code);
  moduleInstance.loaded = true;
  return moduleInstance;
}

/*
 * =====================================
 * Module class used internally and to
 * mock the native Node `module` with
 * exposed methods.
 * =====================================
 */

export class Module {
  id: string;
  filename: string;
  loaded: boolean = false;
  exports: Exports = {};
  parent: ?Module;
  paths: string[] = [];
  children: Module[] = [];
  sourceMap: ?Object;

  constructor(filename: string, parent: ?Module) {
    this.id = filename;
    this.filename = filename;
    this.parent = parent;
    if (this.parent) {
      this.parent.children.push(this);
    }
  }

  /**
   * Transpile and eval the module.
   */
  _compile(code: string, altFilename?: string): Exports {
    const filename = altFilename || this.filename;

    if (/\.json$/.test(filename)) {
      try {
        this.exports = JSON.parse(code);
      } catch (error) {
        throw new Error(`${error.message} (${filename})`);
      }

      return this.exports;
    }

    // Transpile module implementation.
    const { code: moduleBody, map } = babel.transform(code, {
      plugins: [
        require.resolve('babel-plugin-transform-es2015-modules-commonjs'),
      ],
      filename,
      sourceMaps: true,
      ignore: /node_modules/,
    });

    this.paths = NativeModule._nodeModulePaths(path.dirname(filename));
    this.sourceMap = map;

    // Load module.
    try {
      // Create script object with module wrapper.
      const script = new vm.Script(ModuleMock.wrap(moduleBody), {
        filename,
      });

      // Run the script to get the wrapper function.
      const loader = script.runInThisContext();

      // Compile the module with custom module system.
      loader(
        this.exports,
        getRequireMock(this),
        this,
        filename,
        path.dirname(filename)
      );
    } catch (error) {
      if (error.isEnhanced) {
        throw buildCodeFrameError(error, error.enhancedFrames);
      } else {
        let errorToThrow;
        try {
          errorToThrow = buildCodeFrameError(
            error,
            enhanceFrames(
              getFramesFromStack(error, frames =>
                frames.findIndex(frame => frame.fileName === filename)
              ),
              modulesCache
            )
          );
        } catch (_) {
          errorToThrow = error;
        }
        throw errorToThrow;
      }
    }

    return this.exports;
  }

  load(altFilename?: string) {
    const filename = altFilename || this.filename;
    let extension = path.extname(filename) || '.js';
    if (!ModuleMock._extensions[extension]) {
      extension = '.js';
    }
    ModuleMock._extensions[extension](this, filename);
    // Toggle loaded flag.
    this.loaded = true;
  }

  require(moduleId: string): Exports {
    return getRequireMock(this)(moduleId);
  }
}

/*
 * =====================================
 * Native Node `module` mock.
 * =====================================
 */

export function ModuleMock(filename: string, parent: ?Module): Module {
  return new Module(filename, parent);
}

// Copy all function from native node `module`.
Object.keys(NativeModule).forEach(key => {
  ModuleMock[key] = NativeModule[key];
});

ModuleMock.wrap = code =>
  `${NativeModule.wrapper[0]}${code}${NativeModule.wrapper[1]}`;
ModuleMock._cache = modulesCache;
ModuleMock._debug = () => {}; // noop
ModuleMock._load = (moduleId: string, parent: ?Module) =>
  getRequireMock(parent)(moduleId);
ModuleMock._preloadModules = () => {}; // noop
ModuleMock.Module = ModuleMock;

const ModuleMockId = resolveModuleId('module');
modulesCache[ModuleMockId] = new Module(ModuleMockId);
modulesCache[ModuleMockId].loaded = true;
modulesCache[ModuleMockId].exports = ModuleMock;

/*
 * =====================================
 * Require and resolve mocks.
 * =====================================
 */

/**
 * Get mocked require function for specific parent,
 * so that it can resolve other modules relative to
 * parent module.
 */
function getRequireMock(parent: ?Module) {
  function resolveMock(moduleId: string): string {
    return resolveModuleId(moduleId, parent);
  }

  function requireMock(moduleId: string): Exports {
    /**
     * For non JS/JSON requires, we create a dummy wrapper module and just export
     * the moduleId from it, thus letting the bundler handle the rest.
     */
    if (/\.(?!js)[a-zA-Z0-9]+$/.test(moduleId)) {
      return instantiateModule(
        `module.exports = '${moduleId}'`,
        moduleId,
        parent
      ).exports;
    }

    const filename = resolveMock(moduleId);

    // Native Node modules
    if (filename === moduleId && !path.isAbsolute(moduleId)) {
      // $FlowFixMe
      return require(moduleId); // eslint-disable-line global-require
    }

    // Return cached module if available.
    if (modulesCache[filename]) {
      return modulesCache[filename].exports;
    }

    const code = fs.readFileSync(filename, 'utf-8');
    return instantiateModule(code, filename, parent).exports;
  }

  // Provide utilities on require function.
  requireMock.resolve = resolveMock;
  // In our case main will never be set to anything other than undefined,
  // because we never run js module directly with `node module.js`.
  requireMock.main = undefined;
  // $FlowFixMe
  requireMock.extensions = require.extensions;
  requireMock.cache = modulesCache;

  return requireMock;
}

/*
 * =====================================
 * Utils for modules cache manipulation.
 * =====================================
 */

/**
 * Get module instance from cache.
 */
export function getCachedModule(moduleId: string): Module {
  return modulesCache[moduleId];
}

/**
 * Clear modules from cache which are neither from node_modules nor from linaria.
 */
export function clearLocalModulesFromCache() {
  Object.keys(modulesCache)
    .filter(
      moduleId =>
        !/node_modules/.test(moduleId) && !/linaria\/(build|src)/.test(moduleId)
    )
    .forEach(moduleId => {
      delete modulesCache[moduleId];
    });
}

export function clearModulesCache() {
  Object.keys(modulesCache).forEach(moduleId => {
    delete modulesCache[moduleId];
  });
}
