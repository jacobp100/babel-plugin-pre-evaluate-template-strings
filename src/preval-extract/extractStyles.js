/* @flow */

import fs from 'fs';
import path from 'path';
import mkdirp from 'mkdirp';

import type { BabelTypes, NodePath } from '../types';

import { getCachedModule } from '../lib/moduleSystem';
import { relativeToCwd, makeAbsolute } from './utils';

const withPreamble = data =>
  `/* THIS FILE IS AUTOGENERATED. DO NOT EDIT IT DIRECTLY, NOR COMMIT IT TO VERSION CONTROL. */\n${data}`;

/**
 * Get output filename with directory structure from source preserved inside
 * custom outDir.
 */
function getOutputFilename(
  relativeFilename: string,
  absOutDir: string
): string {
  const basename = /(.+)\..+$/.exec(path.basename(relativeFilename))[1];
  const relativeOutputFilename = path.join(
    path.dirname(relativeFilename),
    `${basename}.css`
  );
  return path.join(absOutDir, relativeOutputFilename);
}

let stylesCache = {};

function hasCachedStyles(filename: string, styles: string) {
  return stylesCache[filename] && stylesCache[filename] === styles;
}

function createCssFromCache() {
  return withPreamble(
    Object.keys(stylesCache).reduce(
      (acc, file) => `${acc}\n${stylesCache[file]}`,
      ''
    )
  );
}

function addRequireForCss(
  types: BabelTypes,
  program: NodePath<*>,
  filename: string
) {
  program.node.body.unshift(
    types.expressionStatement(
      types.callExpression(types.identifier('require'), [
        types.stringLiteral(filename),
      ])
    )
  );
}

/**
 * Write styles to file and create directory if needed.
 */
function outputStylesToFile(
  filename: string,
  styles: string,
  throwImmediately: boolean = false
) {
  try {
    fs.writeFileSync(filename, styles);
  } catch (error) {
    if (
      !throwImmediately &&
      (error.code === 'ENOENT' || /ENOENT/.test(error.message))
    ) {
      mkdirp.sync(path.dirname(filename));
      outputStylesToFile(filename, styles, true);
    } else {
      throw error;
    }
  }
}

export default function extractStyles(
  types: BabelTypes,
  program: NodePath<*>,
  currentFilename: string,
  options: {
    single?: boolean,
    filename?: string,
    outDir?: string,
    cache?: boolean,
    extract?: boolean,
  }
) {
  // Normalize current filename path.
  const relativeCurrentFilename = relativeToCwd(currentFilename);
  const absCurrentFilename = makeAbsolute(currentFilename);

  const {
    single = false,
    cache = true,
    extract = true,
    outDir = '.linaria-cache',
    filename: basename = 'styles.css',
  } = options;

  const absOutDir = path.isAbsolute(outDir)
    ? outDir
    : path.join(process.cwd(), outDir);

  // If single === true, we compute filename from outDir and filename options,
  // since there will be only one file, otherwise we need to reconstruct directory
  // structure inside outDir. In that case filename option is discard.
  const filename = single
    ? path.join(absOutDir, basename)
    : getOutputFilename(relativeCurrentFilename, absOutDir);
  const importPath = `./${path.relative(
    path.dirname(absCurrentFilename),
    filename
  )}`;

  if (!extract) {
    return;
  }

  const sheet = getCachedModule(require.resolve('../../sheet.js'));
  const data = sheet ? sheet.exports.default.dump() : '';

  if (!data.length) {
    return;
  }

  if (cache) {
    if (single) {
      // If single === true, we cannot rely on filename since it will
      // always be the same, so we need to use absCurrentFilename.
      if (hasCachedStyles(absCurrentFilename, data)) {
        return;
      }
      stylesCache[absCurrentFilename] = data;
    } else {
      if (hasCachedStyles(filename, data)) {
        addRequireForCss(types, program, importPath);
        return;
      }
      stylesCache[filename] = data;
    }
  }

  if (single) {
    outputStylesToFile(
      filename,
      cache ? createCssFromCache() : withPreamble(data)
    );
  } else {
    outputStylesToFile(filename, withPreamble(data));
    addRequireForCss(types, program, importPath);
  }
}

export function clearCache() {
  stylesCache = {};
}
