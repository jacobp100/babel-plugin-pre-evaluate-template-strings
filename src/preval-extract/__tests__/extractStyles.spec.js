import path from 'path';
import fs from 'fs';
import mkdirp from 'mkdirp';
import { types } from 'babel-core';
import extractStyles, { clearCache } from '../extractStyles';
import { getCachedModule } from '../../lib/moduleSystem';

jest.mock('../../lib/moduleSystem');
jest.mock('fs');
jest.mock('mkdirp');

const getCachedModuleImpl = returnValue => () => ({
  exports: {
    default: {
      dump() {
        return returnValue;
      },
    },
  },
});

describe('preval-extract/extractStyles module', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('should not do anything if there is no styles', () => {
    getCachedModule.mockImplementationOnce(getCachedModuleImpl(''));

    extractStyles(types, null, 'test.js', {});

    getCachedModule.mockImplementationOnce(getCachedModuleImpl(''));
    extractStyles(types, null, 'test.js', {});

    expect(fs.writeFileSync).not.toHaveBeenCalled();
  });

  it('should not do anything if sheet module is not found in cache', () => {
    getCachedModule.mockImplementationOnce(() => null);

    extractStyles(types, null, 'test.js', {});

    expect(fs.writeFileSync).not.toHaveBeenCalled();
  });

  it('should skip extraction if extract option is false', () => {
    getCachedModule.mockImplementationOnce(getCachedModuleImpl('...'));
    const writeFileSync = jest.fn();

    extractStyles(
      types,
      null,
      'filename.js',
      { extract: false },
      { writeFileSync }
    );

    expect(writeFileSync).not.toHaveBeenCalled();
  });

  it('should create directory structure if needed', () => {
    getCachedModule.mockImplementationOnce(
      getCachedModuleImpl('.classname{color: #ffffff}')
    );

    let calls = 0;
    fs.writeFileSync.mockImplementation(() => {
      if (calls === 0) {
        calls++;
        const error = new Error('ENOENT');
        error.code = 'ENOENT';
        throw error;
      }
    });

    const program = {
      node: {
        body: [],
      },
    };

    extractStyles(types, program, 'filename.js', {
      cache: false,
    });

    expect(mkdirp.sync).toHaveBeenCalledTimes(1);
    expect(mkdirp.sync).toHaveBeenCalledWith(
      path.join(process.cwd(), '.linaria-cache')
    );
  });

  it('should throw error if cannot write to file', () => {
    getCachedModule.mockImplementationOnce(
      getCachedModuleImpl('.classname{color: #ffffff}')
    );

    fs.writeFileSync.mockImplementation(() => {
      const error = new Error('ERR');
      error.code = 'ERR';
      throw error;
    });

    const program = {
      node: {
        body: [],
      },
    };

    expect(() => {
      extractStyles(types, program, 'filename.js', {
        cache: false,
      });
    }).toThrowError();
  });

  describe('with cache disabled', () => {
    beforeEach(clearCache);

    it('with single set to true, should write styles to a file', () => {
      getCachedModule.mockImplementationOnce(
        getCachedModuleImpl('.classname{color: #ffffff}')
      );

      const program = {
        node: {
          body: [],
        },
      };
      extractStyles(types, program, 'filename.js', {
        single: true,
        cache: false,
      });

      expect(fs.writeFileSync).toHaveBeenCalledTimes(1);
      expect(fs.writeFileSync.mock.calls[0][0]).toEqual(
        path.join(process.cwd(), '.linaria-cache/styles.css')
      );
      expect(fs.writeFileSync.mock.calls[0][1]).toMatchSnapshot();
      expect(program.node.body.length).toBe(0);
    });

    it('should write (and overwrite) new styles to a file', () => {
      getCachedModule.mockImplementationOnce(
        getCachedModuleImpl('.classname{color: #ffffff}')
      );

      const program = {
        node: {
          body: [],
        },
      };

      extractStyles(types, program, path.join(process.cwd(), 'filename.js'), {
        cache: false,
      });

      getCachedModule.mockImplementationOnce(
        getCachedModuleImpl('.classname{color: #000000}')
      );
      // On each transpilation the body does not contain require call added by extractStyles
      program.node.body = [];

      extractStyles(types, program, path.join(process.cwd(), 'filename.js'), {
        cache: false,
      });

      expect(fs.writeFileSync).toHaveBeenCalledTimes(2);
      expect(fs.writeFileSync.mock.calls[0][0]).toEqual(
        path.join(process.cwd(), '.linaria-cache/filename.css')
      );
      expect(fs.writeFileSync.mock.calls[0][1]).toMatch(
        '.classname{color: #ffffff}'
      );
      expect(fs.writeFileSync.mock.calls[1][0]).toEqual(
        path.join(process.cwd(), '.linaria-cache/filename.css')
      );
      expect(fs.writeFileSync.mock.calls[1][1]).toMatch(
        '.classname{color: #000000}'
      );
      expect(program.node.body.length).toBe(1);
    });
  });

  describe('with cache enabled', () => {
    beforeEach(clearCache);

    it('should write only once if the styles has not changed', () => {
      getCachedModule.mockImplementationOnce(
        getCachedModuleImpl('.classname{color: #ffffff}')
      );

      const program = {
        node: {
          body: [],
        },
      };

      extractStyles(types, program, 'filename.js', { filename: 'A.css' });

      getCachedModule.mockImplementationOnce(
        getCachedModuleImpl('.classname{color: #ffffff}')
      );
      // On each transpilation the body does not contain require call added by extractStyles
      program.node.body = [];

      extractStyles(types, program, 'filename.js', { filename: 'A.css' });

      expect(fs.writeFileSync).toHaveBeenCalledTimes(1);
      expect(fs.writeFileSync.mock.calls[0][0]).toEqual(
        path.join(process.cwd(), '.linaria-cache/filename.css')
      );
      expect(fs.writeFileSync.mock.calls[0][1]).toMatch(
        '.classname{color: #ffffff}'
      );
      expect(program.node.body.length).toBe(1);
    });

    it('should append classnames when writing to single file', () => {
      const opts = { single: true, filename: 'A.css' };

      getCachedModule.mockImplementationOnce(
        getCachedModuleImpl('.classname{color: #ffffff}')
      );

      extractStyles(types, null, 'file1.js', opts);

      getCachedModule.mockImplementationOnce(
        getCachedModuleImpl('.other{color: #000000}')
      );

      extractStyles(types, null, 'file2.js', opts);

      expect(fs.writeFileSync).toHaveBeenCalledTimes(2);
      expect(fs.writeFileSync.mock.calls[0][0]).toEqual(
        path.join(process.cwd(), '.linaria-cache/A.css')
      );
      expect(fs.writeFileSync.mock.calls[0][1]).toMatch(
        '.classname{color: #ffffff}'
      );
      expect(fs.writeFileSync.mock.calls[1][0]).toEqual(
        path.join(process.cwd(), '.linaria-cache/A.css')
      );
      expect(fs.writeFileSync.mock.calls[1][1]).toMatch(
        '.classname{color: #ffffff}\n.other{color: #000000}'
      );
    });

    it('should overwrite if the styles has changed', () => {
      getCachedModule.mockImplementationOnce(
        getCachedModuleImpl('.classname{color: #ffffff}')
      );

      const program = {
        node: {
          body: [],
        },
      };

      const opts = { outDir: 'out-dir' };

      extractStyles(types, program, 'filename.js', opts);

      getCachedModule.mockImplementationOnce(
        getCachedModuleImpl('.classname{color: #000000}')
      );
      // On each transpilation the body does not contain require call added by extractStyles
      program.node.body = [];

      extractStyles(types, program, 'filename.js', opts);

      expect(fs.writeFileSync).toHaveBeenCalledTimes(2);
      expect(fs.writeFileSync.mock.calls[0][0]).toEqual(
        path.join(process.cwd(), 'out-dir/filename.css')
      );
      expect(fs.writeFileSync.mock.calls[0][1]).toMatch(
        '.classname{color: #ffffff}'
      );
      expect(fs.writeFileSync.mock.calls[1][0]).toEqual(
        path.join(process.cwd(), 'out-dir/filename.css')
      );
      expect(fs.writeFileSync.mock.calls[1][1]).toMatch(
        '.classname{color: #000000}'
      );
      expect(program.node.body.length).toBe(1);
    });

    it('should append styles only once if the styles has not changed', () => {
      getCachedModule.mockImplementationOnce(
        getCachedModuleImpl('.classname{color: #ffffff}')
      );

      const opts = { single: true };

      extractStyles(types, null, 'filename.js', opts);

      getCachedModule.mockImplementationOnce(
        getCachedModuleImpl('.classname{color: #ffffff}')
      );

      extractStyles(types, null, 'filename.js', opts);

      expect(fs.writeFileSync).toHaveBeenCalledTimes(1);
      expect(fs.writeFileSync.mock.calls[0][0]).toEqual(
        path.join(process.cwd(), '.linaria-cache/styles.css')
      );
      expect(fs.writeFileSync.mock.calls[0][1]).toMatch(
        '.classname{color: #ffffff}'
      );
    });

    it('should overwrite css if the styles has changed', () => {
      getCachedModule.mockImplementationOnce(
        getCachedModuleImpl('.classname{color: #ffffff}')
      );

      const opts = { single: true };

      extractStyles(types, null, 'filename.js', opts);

      getCachedModule.mockImplementationOnce(
        getCachedModuleImpl('.classname{color: #000000}')
      );

      extractStyles(types, null, 'filename.js', opts);

      expect(fs.writeFileSync).toHaveBeenCalledTimes(2);
      expect(fs.writeFileSync.mock.calls[0][0]).toEqual(
        path.join(process.cwd(), '.linaria-cache/styles.css')
      );
      expect(fs.writeFileSync.mock.calls[0][1]).toMatch(
        '.classname{color: #ffffff}'
      );
      expect(fs.writeFileSync.mock.calls[1][0]).toEqual(
        path.join(process.cwd(), '.linaria-cache/styles.css')
      );
      expect(fs.writeFileSync.mock.calls[1][1]).toMatch(
        '.classname{color: #000000}'
      );
      expect(fs.writeFileSync.mock.calls[1][1]).not.toMatch(
        '.classname{color: #ffffff}'
      );
    });

    it('should overwrite css if the new file was added', () => {
      getCachedModule.mockImplementationOnce(
        getCachedModuleImpl('.classname{color: #ffffff}')
      );

      const opts = { single: true };

      extractStyles(types, null, 'filename1.js', opts);

      getCachedModule.mockImplementationOnce(
        getCachedModuleImpl('.classname{color: #000000}')
      );

      extractStyles(types, null, 'filename2.js', opts);

      expect(fs.writeFileSync).toHaveBeenCalledTimes(2);
      expect(fs.writeFileSync.mock.calls[0][0]).toEqual(
        path.join(process.cwd(), '.linaria-cache/styles.css')
      );
      expect(fs.writeFileSync.mock.calls[0][1]).toMatch(
        '.classname{color: #ffffff}'
      );
      expect(fs.writeFileSync.mock.calls[1][0]).toEqual(
        path.join(process.cwd(), '.linaria-cache/styles.css')
      );
      expect(fs.writeFileSync.mock.calls[1][1]).toMatch(
        '.classname{color: #ffffff}'
      );
      expect(fs.writeFileSync.mock.calls[1][1]).toMatch(
        '.classname{color: #000000}'
      );
    });
  });
});
