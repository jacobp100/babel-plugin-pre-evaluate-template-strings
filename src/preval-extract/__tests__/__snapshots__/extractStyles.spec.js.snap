// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preval-extract/extractStyles module with cache disabled with single set to true, should write styles to a file 1`] = `
"/* THIS FILE IS AUTOGENERATED. DO NOT EDIT IT DIRECTLY, NOR COMMIT IT TO VERSION CONTROL. */
.classname{color: #ffffff}"
`;
